torch/_export/__init__.py

```diff
@@ -559,6 +559,88 @@ def export(
         preserve_module_call_signature=preserve_module_call_signature,
     )
 
+
+def _prepare_module(
+    gm_torch_level: torch.fx.GraphModule,
+    aot_export_args
+) -> List[str]:
+    flat_args = pytree.tree_leaves(aot_export_args)
+    user_input_names = []
+    with gm_torch_level.graph.inserting_before():
+        for i, (arg, node) in enumerate(zip(flat_args, gm_torch_level.graph.nodes)):
+            assert node.op == "placeholder"
+            user_input_names.append(node.name)
+            if isinstance(arg, torch.Tensor):
+                assert not hasattr(gm_torch_level, node.name)
+                gm_torch_level.register_buffer(node.name, arg)
```

Comment:
Hmmm, what will happen in these two cases?

(1) Model has a forward input that requires_grad=True, and passes that input to a triton kernel

(2) Model has a parameter, and passes it to a triton kernel

Reply:
@bdhirsh sorry I just saw your comment today. I was planning to play around triton kernels today but was distracted. Here is what I tried so far:
1. Passing a forward input with requires_grad=True and mutate it inside forward. TorchDynamo actually errors out before we enter aot_export stage with TorchRuntimeError.
2. Added parameter and mutates it inside forward function. TorchDynamo also errors on that with TorchRuntimeError.

Reply:
Circling back on this:
for case (1) you mentioned I got this error:
```
RuntimeError: Found a graph input that requires gradients, and received a mutation.
This is currently banned in the aot_export workflow. If you need this functionality, please file a github issue.
```
for case (2):
```
RuntimeError: Found a graph input that requires gradients, and received a mutation.
This is currently banned in the aot_export workflow. If you need this functionality, please file a github issue.
```
These are expected from my point of view.

test case I used:
```
    def test_export_input_mutation_static_shape_triton(self):
        from torch.testing._internal.triton_utils import (
            add_kernel,
        )
        import triton

        class MutationModelTriton(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.param = torch.nn.Parameter(torch.randn(12))

            def forward(self, x, y):
                n_elements = y.numel()
                grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
                add_kernel[grid](self.param, x, y, n_elements, BLOCK_SIZE=16)
                return x, y

        inputs = (torch.randn(12, device="cuda"), torch.zeros(12, device="cuda", requires_grad=True))
        model = MutationModelTriton()
        ep = torch.export.export(model, inputs)

```

Reply:
> These are expected from my point of view.

Oh are we ok with these being errors? My assumption was that people might end up trying to export models that pass arbitrary tensors into triton kernels (e.g. weights or inputs), but I could be wrong.

Reply:
@bdhirsh Yes we're ok with these being errors. The rationale is, if user saw error like this when exporting a model, it will be very easy for them to change inputs to requires_grad=False without changing modeling or exporting code (most of the cases, people only want forward graph anyway). Also, even for normal aten op we don't allow this in dynamo (my first comment), so I think triton's behavior looks consistent here actually. 

The error message itself subjectively could be improved tho lol, like "hint: triton kernel will assume every input is mutated". I'm not very good at error messaging but I think the error itself looks reasonable enough to me.